blockcache/blockcache.go

```diff
@@ -0,0 +1,224 @@
+}
+
+// MostRecentBlockCache is an implementation of the BlockCache interface
+// which attempts to cache a fixed number of the most recent blocks.
+// If a newer block (by height) is inserted and the cache is already full,
+// then the oldest block will be evicted to make room.
+type MostRecentBlockCache struct {
+	db *bolt.DB
+
+	// A BlockHeaderStore allows translation between block heights and hashes.
+	blockHeaders *headerfs.BlockHeaderStore
+
+	// The maximum number of blocks to store in the cache.
+	capacity int
+
+	// The number of cache hits and misses. Must be accessed atomically.
+	hits   uint64
+	misses uint64
+}
+
+// A compile-time check to ensure the MostRecentBlockCache adheres to the
+// BlockCache interface.
+var _ BlockCache = (*MostRecentBlockCache)(nil)
+
+// fileExists returns true if the file exists, and false otherwise.
+func fileExists(path string) bool {
+	if _, err := os.Stat(path); err != nil {
+		if os.IsNotExist(err) {
+			return false
+		}
+	}
+
+	return true
+}
+
+// New creates a new instance of a MostRecentBlockCache given a path to store
+// the data.
+func New(dbPath string, capacity int, blockHeaders *headerfs.BlockHeaderStore) (*MostRecentBlockCache, error) {
+	if !fileExists(dbPath) {
+		if err := os.MkdirAll(dbPath, 0700); err != nil {
+			return nil, err
+		}
+	}
+
+	path := filepath.Join(dbPath, dbName)
+	bdb, err := bolt.Open(path, dbFilePermission, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	// Create the bucket if necessary.
+	err = bdb.Update(func(tx *bolt.Tx) error {
+		_, err := tx.CreateBucket(blockCacheBucket)
+
+		if err != nil && err != bolt.ErrBucketExists {
+			return err
+		}
+
+		return nil
+	})
+
+	if err != nil {
+		return nil, fmt.Errorf("unable to create cache db")
+	}
+
+	return &MostRecentBlockCache{
+		db:           bdb,
+		capacity:     capacity,
+		blockHeaders: blockHeaders,
+	}, nil
+}
+
+// Close instructs the underlying database to stop cleanly.
+func (c *MostRecentBlockCache) Close() {
+	c.db.Close()
+}
+
+// PutBlock attempts to insert the block into the cache. If the cache is at
+// capacity it will first attempt to evict the oldest block, if the given block
+// is older than all blocks currently in the cache, then no action happens.
+func (c *MostRecentBlockCache) PutBlock(block *wire.MsgBlock) error {
+	// The key for this block is its height in the blockchain.
+	// This allows us to evict the oldest entry easily.
+	hash := block.BlockHash()
+	height, err := c.blockHeaders.HeightFromHash(&hash)
+	if err != nil {
+		return err
+	}
+
+	key := []byte(strconv.FormatUint(uint64(height), 10))
+
+	// Serialize the block.
+	w := bytes.NewBuffer(make([]byte, 0, block.SerializeSize()))
+	err = block.Serialize(w)
+	if err != nil {
+		return err
+	}
+
+	// Insert the block, evicting the oldest block if the capacity has been
+	// reached.
+	err = c.db.Update(func(tx *bolt.Tx) error {
+		bucket := tx.Bucket(blockCacheBucket)
+
+		isFull := bucket.Stats().KeyN >= c.capacity
```

Comment:
If moving to size based, this can be removed (and also the logic below modified a bit).